agent/agent.go

```diff
@@ -599,98 +599,96 @@ func formatReportLine(cells ...string) string {
 	return fmt.Sprintf(format, strValues...)
 }
 
-// TODO(mkcp): This duplicates much of customOps and can certainly be improved.
-func customHostRunners(cfg *HostConfig, tmpDir string) ([]runner.Runner, error) {
+// TODO(mkcp): Products, not the agent, should handle their own custom ops when they're created.
+func customRunners[C HCLConfig](config C, tmpDir string) ([]runner.Runner, error) {
 	runners := make([]runner.Runner, 0)
-	// Build Commanders
-	for _, c := range cfg.Commands {
-		cmder := runner.NewCommander(c.Run, c.Format)
-		runners = append(runners, cmder)
-	}
-	// Build Shellers
-	for _, c := range cfg.Shells {
-		sheller := runner.NewSheller(c.Run)
-		runners = append(runners, sheller)
-	}
-
-	for _, g := range cfg.GETs {
-		runners = append(runners, host.NewGetter(g.Path))
-	}
+	switch cfg := any(config).(type) {
+	case *ProductConfig:
```

Comment:
I notice that this function is now ~90 lines long and looks like it has several branches along the way. What if the new "genericized" function had the type switch, but then simply called out to non-generic functions based on the type? Essentially, the same two functions as before would exist and could be private/unexported functions, but this generic one would be the entrypoint? In prepping to split out a library, perhaps it could also be an exported function? It would still allow for a simpler interface for callers, I think, but allow us to test the sub-functions more easily.

Reply:
I ended up solving this problem by treating `customRunners()` as a reducer from `HCLConfig` to `[]Runner`. There's now mapping functions that correspond to each kind of RunnerConfig, like `[]ShellConfig`, and each applies the config structs to the corresponding `runner.New<Runner>()` constructor. This trims down `customRunners()` quite a bit and matches the shape of the data by destructuring each field in the `HCLConfig` types to their corresponding mapper.